src/deepsparse/transformers/evaluator.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from types import SimpleNamespace
+from typing import Any, Dict, List, Optional
+
+from deepsparse.pipeline import DEEPSPARSE_ENGINE
+from deepsparse.transformers.eval_downstream import perplexity_eval
+
+
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.INFO)
+
+
+@dataclass
+class Evaluator(ABC):
+    """
+    Evaluator base class that represents the contract all evaluators must follow.
+
+    :param model: The path to the model to evaluate.
+    :param dataset: The name of the dataset to evaluate on.
+    :param eval: The list of evaluation metrics to run.
+    """
+
+    def __init__(self, model: str, dataset: str, eval: Optional[List[str]] = None):
+        self.model_path_ = model
+        self.dataset_name_ = dataset
+        self.eval_ = eval
+        self.results_ = None
+
+    @abstractmethod
+    def evaluate(self, *args, **kwargs):
+        """
+        Main method for running the evaluation.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_results(self):
+        """
+        Return the results of the evaluation.
+        """
+        raise NotImplementedError
+
+
+class LMEvaluator(Evaluator):
```

Comment:
LM stands for Language Models? So we can evalute both "base" transformers as well as LLMs?

Reply:
Yes Perplexity can be calculated for all LMs

Reply:
I am just not sure about the term LM. May not be obvious what this means. Can we just rename it to TransformersEvaluator or NLPEvaluator?

Reply:
Do you think PerplexityEvaluator is better? as this does not work with all Transformers/all NLP models?

Reply:
Perplexity works for all the Transformer models as far as I know. But the cool thing about your evaluator is that you can extend it to other "LLM-type" metrics like pass@k.

Reply:
`TransformersEvaluator` sounds like the best choice for me imo.

Reply:
Accepted!